### PR TITLE
devicetree: edtlib: gen_defines: Add model name macros based on compatibles

### PR DESCRIPTION
**Pull request by MaureenHelm**
Follow up to #49489 adding support in edtlib, gen_defines, and devicetree.h to get model name strings from a node's compatibles.

cc: @huhebo

### CODE REVIEW

**Review by yperess — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

File: scripts/dts/python-devicetree/src/devicetree/edtlib.py [hunk 1]
```diff
@@ -477,9 +477,9 @@ def _init_luts(self):
                          f"'{compat_re}'")
 
                 if ',' in compat and self._vendor_prefixes:
-                    vendor = compat.split(',', 1)[0]
+                    vendor, model = compat.split(',', 1)
                     if vendor in self._vendor_prefixes:
-                        self.compat2vendor[compat] = self._vendor_prefixes[vendor]
+                        self.compat2vendor[compat] = (self._vendor_prefixes[vendor], model)
```
Comment (mbolivar-nordic) [inline scripts/dts/python-devicetree/src/devicetree/edtlib.py:482]: Please rename the variable or create a separate look up table.

File: include/zephyr/devicetree.h [hunk 2]
```diff
@@ -1796,6 +1796,82 @@
...
+ * Notice that the compatible at index 1 doesn't match any entries in the
+ * vendor prefix file and therefore index 1 is not a valid model index. Use
+ * DT_NODE_MODEL_HAS_IDX(node_id, idx) to determine if an index is valid.
+ *
+ * @param node_id node identifier
+ * @param idx index of the model to return
+ * @return string literal of the idx-th model
+ */
+#define DT_NODE_MODEL_BY_IDX(node_id, idx) \
+	DT_CAT(node_id, _COMPAT_MODEL_IDX_##idx)
```
Comment (mbolivar-nordic) [inline include/zephyr/devicetree.h:1829]: Use DT_CAT3 (see 47ef58e8ac4aecc186054cf703a0a62a0106957f).

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

**Comment by galak** [comment]
Hmm, curious about if we think a compatible w/o vendor means, no model.

For example:

`compatible = "ns16550";`

seems like model being `ns16550` would be reasonable.

@mbolivar-nordic @MaureenHelm thoughts?

**Comment by mbolivar-nordic** [comment]
@galak do you really mean to say "no model"? I'm confused because your example seems to be saying the opposite: compatible without vendor means the compatible is the model, not compatible with no vendor means no model.

Are you asking for a change?

**Comment by galak** [comment]
> @galak do you really mean to say "no model"? I'm confused because your example seems to be saying the opposite: compatible without vendor means the compatible is the model, not compatible with no vendor means no model.
> 
> Are you asking for a change?

I'm asking for us to decide how to handle the case of a `compatible` that is NOT of the form `vendor,model`.  For `compatible = "ns16550"`, is the `model` None or is it `ns16550`?

**Comment by MaureenHelm** [comment]
> > @galak do you really mean to say "no model"? I'm confused because your example seems to be saying the opposite: compatible without vendor means the compatible is the model, not compatible with no vendor means no model.
> > Are you asking for a change?
> 
> I'm asking for us to decide how to handle the case of a `compatible` that is NOT of the form `vendor,model`. For `compatible = "ns16550"`, is the `model` None or is it `ns16550`?

I took the simplest route and chose not to define a model macro when the compatible doesn't have the form `vendor,model`. The logic is consistent with the vendor macro, which also doesn't exist for non-compliant compatibles. I'm open to other ideas though.

**Comment by galak** [comment]
> > > @galak do you really mean to say "no model"? I'm confused because your example seems to be saying the opposite: compatible without vendor means the compatible is the model, not compatible with no vendor means no model.
> > > Are you asking for a change?
> > 
> > 
> > I'm asking for us to decide how to handle the case of a `compatible` that is NOT of the form `vendor,model`. For `compatible = "ns16550"`, is the `model` None or is it `ns16550`?
> 
> I took the simplest route and chose not to define a model macro when the compatible doesn't have the form `vendor,model`. The logic is consistent with the vendor macro, which also doesn't exist for non-compliant compatibles. I'm open to other ideas though.

For vendor it makes sense, since if there is a `,` than we clearly have no vendor.  My question is that it seems like model still might make sense for the cases that there is NOT `,`.

**Comment by mbolivar-nordic** [comment]
In my opinion we shouldn't define a model in cases like that. Things like `gpio-leds` are not a model of any hardware in the sense I usually understand the term. The absence of a vendor means "generic or vendor-independent" almost all the time, which is why the [bindings index](https://docs.zephyrproject.org/latest/build/dts/api/bindings.html#dt-no-vendor) uses that as the section name when there is no vendor.

**Review by galak — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@mbolivar-nordic are you ok with this now?

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**